sui/src/rpc_gateway.rs

```diff
@@ -0,0 +1,390 @@
+use sui_core::gateway_state::gateway_responses::TransactionResponse;
+use sui_core::gateway_state::{GatewayClient, GatewayState, GatewayTxSeqNumber};
+use sui_types::base_types::{ObjectID, SuiAddress, TransactionDigest};
+use sui_types::crypto;
+use sui_types::crypto::SignableBytes;
+use sui_types::messages::{Transaction, TransactionData};
+use sui_types::object::ObjectRead;
+
+use crate::config::PersistedConfig;
+use crate::gateway::GatewayConfig;
+use crate::rest_gateway::responses::{NamedObjectRef, ObjectResponse};
+
+#[rpc(server, client, namespace = "sui")]
+pub trait RpcGateway {
+    #[method(name = "getObjectInfo")]
+    async fn get_object_info(&self, object_id: ObjectID) -> RpcResult<ObjectRead>;
+
+    #[method(name = "transferCoin")]
+    async fn transfer_coin(
+        &self,
+        signer: SuiAddress,
+        object_id: ObjectID,
+        gas_payment: ObjectID,
+        gas_budget: u64,
+        recipient: SuiAddress,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "moveCall")]
+    async fn move_call(
+        &self,
+        signer: SuiAddress,
+        package_object_id: ObjectID,
+        module: Identifier,
+        function: Identifier,
+        type_arguments: Vec<TypeTag>,
+        pure_arguments: Vec<Base64EncodedBytes>,
+        gas_object_id: ObjectID,
+        gas_budget: u64,
+        object_arguments: Vec<ObjectID>,
+        shared_object_arguments: Vec<ObjectID>,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "publish")]
+    async fn publish(
+        &self,
+        sender: SuiAddress,
+        compiled_modules: Vec<Base64EncodedBytes>,
+        gas_object_id: ObjectID,
+        gas_budget: u64,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "splitCoin")]
+    async fn split_coin(
+        &self,
+        signer: SuiAddress,
+        coin_object_id: ObjectID,
+        split_amounts: Vec<u64>,
+        gas_payment: ObjectID,
+        gas_budget: u64,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "mergeCoins")]
+    async fn merge_coin(
+        &self,
+        signer: SuiAddress,
+        primary_coin: ObjectID,
+        coin_to_merge: ObjectID,
+        gas_payment: ObjectID,
+        gas_budget: u64,
+    ) -> RpcResult<TransactionBytes>;
+
+    #[method(name = "executeTransaction")]
+    async fn execute_transaction(
+        &self,
+        signed_transaction: SignedTransaction,
+    ) -> RpcResult<TransactionResponse>;
+
+    #[method(name = "syncAccountState")]
+    async fn sync_account_state(&self, address: SuiAddress) -> RpcResult<()>;
+
+    #[method(name = "getOwnedObjects")]
+    async fn get_owned_objects(&self, owner: SuiAddress) -> RpcResult<ObjectResponse>;
+
+    #[method(name = "getTotalTransactionNumber")]
+    async fn get_total_transaction_number(&self) -> RpcResult<u64>;
+
+    #[method(name = "getTransactionsInRange")]
+    async fn get_transactions_in_range(
+        &self,
+        start: GatewayTxSeqNumber,
```

Comment:
How would a client know what the sequence numbers to use here are?

Reply:
> It would be super nice to auto generate the client code, so I really wonder about the support for things like that.

jsonrpsee supports client code generation in rust, we are using that in https://github.com/MystenLabs/sui/blob/main/sui/src/rpc_gateway_client.rs

Reply:
I am not very familiar with the use case but I guess it's for paging? These methods are mirroring GatewayAPI, maybe @lxfind will have more context.